src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2153,59 +2153,58 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
             else if (Vector256.IsHardwareAccelerated && length >= Vector256<TValue>.Count)
             {
                 Vector256<TValue> equals, values = Vector256.Create(value);
-                ref TValue currentSearchSpace = ref Unsafe.Add(ref searchSpace, length - Vector256<TValue>.Count);
+                nint offset = length - Vector256<TValue>.Count;
 
-                // Loop until either we've finished all elements or there's less than a vector's-worth remaining.
-                do
+                // Loop until either we've finished all elements -or- there's one or less than a vector's-worth remaining.
+                while (offset > 0)
                 {
-                    equals = TNegator.NegateIfNeeded(Vector256.Equals(values, Vector256.LoadUnsafe(ref currentSearchSpace)));
+                    equals = TNegator.NegateIfNeeded(Vector256.Equals(values, Vector256.LoadUnsafe(ref searchSpace, (nuint)(offset))));
+
                     if (equals == Vector256<TValue>.Zero)
                     {
-                        currentSearchSpace = ref Unsafe.Subtract(ref currentSearchSpace, Vector256<TValue>.Count);
+                        offset -= Vector256<TValue>.Count;
                         continue;
                     }
 
-                    return ComputeLastIndex(ref searchSpace, ref currentSearchSpace, equals);
+                    return ComputeLastIndex(offset, equals);
                 }
-                while (!Unsafe.IsAddressLessThan(ref currentSearchSpace, ref searchSpace));
```

Comment:
I'm being dense, but I'm still not clear on what the actual problem here was.  I see that `currentSearchSpace` could end up pointing to before the object, but at that point it's never dereferenced.  Other than a "SanityCheck" assert, is there an actual negative impact that ref would have?  If it ended up pointing into some unrelated object, it could root that object unnecessarily, and if that object moved, the value of this pointer could end up changing, but if it does change we're not going to use it again, so that shouldn't be visible. Is just the existence of that ref somehow a real problem?

Reply:
The issue is that once it points to before the object, its no longer tracked and won't be updated if the GC decides to relocate `searchSpace`.

This means that if `searchSpace` is moved to a lower address, the comparison will be `false` (and thus loop again) when it should have been `true`.

Reply:
> Other than a "SanityCheck" assert, is there an actual negative impact that ref would have?

The GC can move the previous object that the byref pointed into to be on the other side the current buffer that we are searching. It would make the `IsAddressLessThan` check succeed, the loop the continue and eventually either crash or find some completely unrelated match.

Reply:
> This means that if searchSpace is moved to a lower address, the comparison will be false (and thus loop again) when it should have been true.

Aha, I see, that makes sense.  It's not that this pointer is a problem in and of itself, it's that the object could be moved to a lower address and this pointer wouldn't be updated, at which point it's pointing past the end of the object.  Got it.

Reply:
The same would be true for the inverse direction, where if `currentSearchSpace` pointed past the end of the object and `searchSpace` were moved to a higher address, the comparison would be incorrect.

Reply:
Also, the GC is not robust against random byrefs that may be pointing into GC owned memory where no managed reference have business in pointing to.

The contract for byrefs is that they have to be either pointing into unmanaged memory that you own, or they are proper managed byrefs. The details were further clarified in https://github.com/dotnet/runtime/pull/71794 . 

It is not valid to take a random number and convert it to byref. It can crash or corrupt the GC.

Reply:
> It is not valid to take a random number and convert it to byref. It can crash or corrupt the GC.

Ok, this is what I thought folks were saying the problem was initially, and I actually didn't realize the GC wasn't robust against that.

Reply:
So, does this mean if I have code like:
```C#
byte* ptr = Marshal.AllocHGlobal(...);
ref byte b = ref Unsafe.AsRef<byte>(ptr);
Use(ref b);
Marshal.FreeHGlobal(ptr);
...
```
that if the JIT extends the lifetime of `b` beyond that FreeHGlobal, it's a bug?

Reply:
I think that would depend on the GC implementation.

Most unmanaged heaps are immovable and are isolated such that the GC could never get an address into that range. Thus, it should never view it as a potential managed byref (something pointing in to a managed heap) and will never try to update it.

Reply:
> that if the JIT extends the lifetime of b beyond that FreeHGlobal, it's a bug?

Yes, it is a potential bug if the GC allocates the memory block that was just freed and starts using it for managed heap.

Reply:
We have a config in GC (`GCConfig::GetConservativeGC`) that enables additional set of validations during stack scanning that are supposed to make the GC robust against the "random" byrefs. It is not heavily tested and probably incomplete. However, we can consider enabling this by default if we think that it is a problem. These checks are relatively cheap (compared to everything else that needs to happen to scan byrefs.)

Reply:
This is an interesting problem. I have opened https://github.com/dotnet/runtime/issues/75865 to have it tracked.

Reply:
I think it would be reasonable to require that the user nulls out the unsafe ref before freeing the memory that it points to.

like:

```cs
byte* ptr = Marshal.AllocHGlobal(...);
ref byte b = ref Unsafe.AsRef<byte>(ptr);
Use(ref b);

// null the ref before releasing the memory
b = ref *(byte*)0;

Marshal.FreeHGlobal(ptr);
```

Once the user started using "Unsafe", they would have to know what they are doing. Especially when representing native memory, not managed by the runtime, as **managed** refs.

Reply:
The JIT is free to extend lifetimes and ignore nulling out of locals like this. Also, it is not unusual for the JIT to create invisible temporary locals that are impossible for the user code to access or null out. We would have to prohibit these optimizations in the JIT for byrefs to make the above fully reliable.

Reply:
But alternative is basically the same as enabling conservative stack scanning. That would break things like unloadability.